File: java/leetcode/RectangleArea.java
//Leetcode problem 223 Rectangle Area
//Leetcode Problem No 850 Rectangle Area II
//Solution written by Xuqiang Fang on 23 April, 2018 
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.HashSet;
import java.util.TreeMap;
import java.util.Set;
import java.util.PriorityQueue;
import java.util.Stack;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Arrays;
class Solution{
    public int computeArea(int A, int B, int C, int D, int E, int F, int G, int H){
        int areaA = (C-A) * (D-B);
        int areaB = (G-E) * (H-F);
        if(C <= E || G <= A || H <= B || D <= F) 
            return areaA + areaB;

        int width = interval(A,C,E,G);
        int height = interval(B,D,F,H);

        return areaA+areaB - width * height;
    }	

    public int interval(int l1, int r1, int l2, int r2){
        if(l1 >= l2 && r1 <= r2)
            return r1-l1;
        else if(l1 <= l2 && r1 >= r2)
            return r2-l2;
        else if(l1 <= l2 && r1 >= l2)
            return r1-l2;
        else
            return r2-l1;
    }
    //No 850
    public int rectangleArea(int[][] rec) {
        int mod = 1000000000+7;
        Map<Integer, Integer> map = new TreeMap<>();
        for(int[] r : rec){
            map.put(r[0], r[0]);
            map.put(r[2], r[2]);
        }
        int index = 0;
        for(int k : map.keySet()){
            map.put(k, index++);
        }
        List<Integer> x = new ArrayList<>(map.keySet());
        int n = x.size();
        int[] count = new int[n];
        List<int[]> L = new ArrayList<>();
        for(int[] r : rec){
            int x1 = r[0], y1 = r[1], x2 = r[2], y2 = r[3];
            L.add(new int[]{y1, x1, x2, 1});
            L.add(new int[]{y2, x1, x2, -1});
        }
        L.sort((a,b)->(a[0]-b[0]));
        long cur_y = 0, cur_x_sum = 0, area = 0;
        for(int[] l : L){
            int y = l[0], x1 = l[1], x2 = l[2];
            int sig = l[3];
            area = (area + (long)(y - cur_y) * (long)(cur_x_sum));
            area %= mod;
            cur_y = y;
            for(int i=map.get(x1); i<map.get(x2); ++i){
                count[i] += sig;
            }
            cur_x_sum = 0;
            for(int i=0; i<n; ++i){
                if(count[i] > 0){
                    cur_x_sum += x.get(i+1) - x.get(i);
                }
            }
        }
        return (int)area;
    }
}

public class RectangleArea{
	public static void main(String[] args){
		Solution s = new Solution();
        //System.out.println(s.computeArea(-3,0,3,4,0,-1,9,2));
        //System.out.println(s.computeArea(-3,0,3,4,0,-1,9,2));
        int[][] rec = {{0,0,2,2},{1,0,2,3},{1,0,3,1}}; 
        System.out.println(s.rectangleArea(rec));
        int[][] rec2 = {{0,0,1000000000,1000000000}};
        System.out.println(s.rectangleArea(rec2));
	}
}

File: cpp/leetcode/RectangleArea.cpp
//Leetcode Problem No 850 Rectangle Area II
//Solution written by Xuqiang Fang on 10 June, 2018

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <stack>
#include <queue>

using namespace std;
class Solution{
public:
    //cannot understand it
    int rectangleArea(vector<vector<int>>& rec) {
        int mod = 1000000000+7;
        vector<int> x{0};
        for(auto r : rec){
            x.push_back(r[0]);
            x.push_back(r[2]);
        }
        sort(x.begin(), x.end());
        vector<int>::iterator end_unique = unique(x.begin(), x.end());
        x.erase(end_unique, x.end());
        unordered_map<int, int> x_i;
        for(int i=0; i<x.size(); ++i){
            //cout << x[i] << endl;
            x_i[x[i]] = i;
        }
        vector<int> count(x.size(), 0);
        vector<vector<int>> L;
        for(auto r : rec){
            int x1 = r[0], y1 = r[1], x2 = r[2], y2 = r[3];
            L.push_back({y1, x1, x2, 1});
            L.push_back({y2, x1, x2, -1});
        }
        sort(L.begin(), L.end());
        long long cur_y = 0, cur_x_sum = 0, area = 0;
        for(auto l : L){
            long long y = l[0], x1 = l[1], x2 = l[2];
            int sig = l[3];
            area = (area + (y - cur_y) * cur_x_sum) % mod;
            cur_y = y;
            for(int i=x_i[x1]; i<x_i[x2]; ++i){
                count[i] += sig;
            }
            cur_x_sum = 0;
            for(int i=0; i<x.size(); ++i){
                if(count[i] > 0){
                    cur_x_sum += x[i+1] - x[i];
                }
            }
        }
        return area;
    }
private:
    long helper(const vector<int>& base, const vector<int>& v){
        long horizontal = overlap(base[0], base[2], v[0], v[2]);
        long vertical = overlap(base[1], base[3], v[1], v[3]);
        //cout << "h = " << horizontal << ", v = " << vertical << endl;
        return horizontal * vertical; 
    }
    long overlap(int x1, int x2, int y1, int y2){
        //x2 >= x1 && y2 >= y1
        if(x1 > y1 && x1 <= y2){
            return (long)min(y2-x1, x2-x1); 
        }
        else if(x1 <= y1){
            if(y2 <= x2){
                return (long)(y2-y1);
            }
            return (long)(x2-y1);
        }
        return 0;
    }
    bool check(int s1, int e1, int s2, int e2){
        return max(s1, s2) < min(e1, e2);
    }
};

int main(){
    Solution s;
    vector<vector<int>> r{{0,0,2,2},{1,0,2,3},{1,0,3,1}};
    cout << s.rectangleArea(r) << endl;
    vector<vector<int>> rec{{0,0,1000000000,1000000000}};
    cout << s.rectangleArea(rec) << endl;
    return 0;
}
